job-info: use job manager journal to track job state

chu11 opened the pull request:
Built on top of #3253.

Per issue #3234, refactor the job-info module to use only the job-manager event stream, not job-state transitions.

chu11 commented:
I should probably mention, this along w/ PR #3253 does technically introduce 1 small racy scenario on `job-info`'s initialization from the KVS.

The events stream is sent out before the KVS eventlog commits are confirmed, so there could be a small window on `job-info`'s loading where data in the KVS eventlog is not complete and an event from the `job-manager.events` stream is missed.  This will be solved w/ journaling support from #3209.

I think this is acceptable in the interim, as its only for when `job-info` is initialized at start, not during normal operations afterwards.

chu11 commented:
re-pushed, noticed a minor cleanup I could do

chu11 commented:
re-pushed, the PR is not too different than before, but it is now built on top of PR #3261.  So some code that was in the old #3261 PR is now in here.

So folks don't have to look it up.  The PR order is:

#3236 job-manager: Move annotations updates to `job-manager.events`, remove `job-annotations` event
#3261 job-manager: support events journal
#3254 job-info: refactor to use only event stream, not job-state transitions

chu11 commented:
re-pushed, think there was a racy test in built-upon PR, rebased while i was at it

chu11 commented:
re-pushed, fixing up given changes in #3236 and #3261.

The major change is that now that this PR depends on #3261 (where as before #3261 depended on this PR), a significant number of changes had to all be absorbed into just one commit (https://github.com/flux-framework/flux-core/pull/3254/commits/64c6d00eb6e64e369f617357f36d6d4a1c08a9e8) to avoid races / tests failing.

chu11 commented:
re-pushed based on changes in #3261

garlick commented:
Ready for a review on this one @chu11?

chu11 commented:
@garlick yup, thanks!

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1745,45 +1677,126 @@ static int job_events_annotations (struct job_state_ctx *jsctx,
 }
 
-static void job_events_continuation (flux_future_t *f, void *arg)
+static int job_events_process (struct job_state_ctx *jsctx, json_t *events)
```
The `job_events_process()` return code is never checked so should it be void?

chu11 replied:
the return code is checked in `job_events_continuation()`

garlick replied:
sorry - missed that

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1745,45 +1677,126 @@ static int job_events_annotations (struct job_state_ctx *jsctx,
-            if (job_events_priority (jsctx, id, timestamp, context) < 0)
-                goto error;
+        if (!strcmp (name, "submit")) {
+            if (job_events_submit (jsctx,
```
The long set of if/elses should probably be broken out of the for loop into a separate function for readability.

Any refactoring possible here?  There's a lot of repetition, and some function calls called here are just wrappers for other functions.

chu11 replied:
> Any refactoring possible here?

I couldn't come up with something simple/obvious here, as many of the function calls are just different enough.  But now that you mention it, I can remove some of the dumb "wrapping" function calls, which might make things look simpler.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1745,45 +1677,126 @@ static int job_events_annotations (struct job_state_ctx *jsctx,
+    return 0;
+}
+
+static void job_events_continuation (flux_future_t *f, void *arg)
```
job_events_continuation() is a continuation of the function `job_state_create()` which is located far away in the source.  Maybe at least rename that to `job_state_create_continuation()` and relocate adjacent to `job_state_create()`?

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1745,45 +1677,126 @@ static int job_events_annotations (struct job_state_ctx *jsctx,
+    }
+    else {
+        if (job_events_process (jsctx, events) < 0)
+            goto error;
```
errno is not set for the `!json_is_array()` case nor the `zlistx_add_end()` failure cases.  The other cases need an audit.

Errno needs to be set before calliing `flux_reactor_stop_error()` so that `flux_reactor_run()`  has that errno set when it returns -1.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1551,26 +1523,24 @@ static int job_events_priority (struct job_state_ctx *jsctx,
-                                search_direction (job));
-        }
+    if (!(job = zhashx_lookup (jsctx->index, &id))) {
+        flux_log_error (jsctx->h, "%s: job %ju not in hash",
```
set errno?

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1659,21 +1609,18 @@ static int job_events_annotations (struct job_state_ctx *jsctx,
-        if (update_early_annotations (jsctx, id, annotations) < 0)
-            return -1;
+    if (!(job = zhashx_lookup (jsctx->index, &id))) {
+        flux_log_error (jsctx->h, "%s: job %ju not in hash",
```
set errno

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -1454,6 +1411,28 @@ static int job_events_alloc (struct job_state_ctx *jsctx,
+        || json_unpack (context, "{ s:i }", "status", &status) < 0) {
+        flux_log (h, LOG_ERR, "%s: finish context invalid: %ju",
+                  __FUNCTION__, (uintmax_t)job->id);
+        return -1;
```
ensure errno gets set since `eventlog_restart_parse()` expects it.  Audit all the other context_parse functions.

garlick reviewed (changes requested):
Found a few errno problems.

chu11 commented on src/modules/job-info/job_state.c:
```diff
@@ -197,22 +192,26 @@ struct job_state_ctx *job_state_create (struct info_ctx *ctx)
-    }
+    zlistx_set_destructor (jsctx->events_journal_backlog, json_decref_wrapper);
 
+    fmt="{s:{s:i s:i s:i s:i s:i s:i s:i s:i s:i s:i}}";
```
duh!  this is suddenly obvious, just not so obvious when you're slowly changing over support 1 event at a time :P

chu11 commented:
rebased & re-pushed, addressing the issues listed above.  Notes on what I did:

- I had forgotten that I wrote a lot of this code before renaming `job-manager.events` to `job-manager.events-journal`, so I renamed a number of functions as a result.  `job_events_continutation()` is now `job_events_journal_continutation()`, many of the "job_events_()` functions are named differently now too (such as `journal_submit_event()`).

- removed the `allow` filter, just accepting all events now.

- audited for errno scenarios.  Hopefully caught all of them.

- When a job does not exist in the primary job hash, I elected to log an error but not return an error leading to `flux_reactor_stop_error()` being called.  It seemed unnecessary to shut down `job-info` in that case.

- regarding the review "Any refactoring possible here? There's a lot of repetition, and some function calls called here are just wrappers for other functions.", I removed a number of the "wrapping" functions that did nothing more than call another function.  That should remove some of the code bloat.

- Regarding the review about "job_events_continuation() is a continuation of the function job_state_create() which is located far away in the source. Maybe at least rename that to job_state_create_continuation() and relocate adjacent to job_state_create()?", I renamed the function per above comment.  Then I did something far simpler.  I simply moved `job_state_create()` and `job_state_destroy()` down to the bottom of the file.  Putting the "context" create/destroy at the bottom of the file is how we do it in a lot of other modules.  I did this in a new commit at the very top of the series.

- there are a couple of new "cleanup" commits at the beginning of the series, setting up fixes later on down the line.

garlick reviewed (approved):
OK, this all looks good, not that my review was incredibly thorough...

You might change the PR title to something like
> job-info: use job manager journal to track job state

That'll be clearer in our release notes.

Feel free to set MWP once you're ready.

chu11 commented:
@mergifyio rebase

chu11 commented:
@garlick cool, thanks!
